src/EditorFeatures/Core/Shared/Extensions/OperationContextExtensions.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Shared.Extensions
+{
+    internal static class OperationContextExtensions
+    {
+        /// <summary>
+        /// Wait until given document is fully loaded and then add scope
+        /// </summary>
+        public static IUIThreadOperationScope WaitForDocumentAndAddScope(
+            this IUIThreadOperationContext context, Document document, bool allowCancellation, string description)
+        {
+            return WaitAndAddScope(context, document.Project.Solution, document.Name, allowCancellation, description);
+        }
+
+        public static IUIThreadOperationScope WaitForProjectAndAddScope(
+            this IUIThreadOperationContext context, Project project, bool allowCancellation, string description)
+        {
+            return WaitAndAddScope(context, project.Solution, project.Name, allowCancellation, description);
+        }
+
+        private static IUIThreadOperationScope WaitAndAddScope(
+            this IUIThreadOperationContext context, Solution solution, string name, bool allowCancellation, string description)
+        {
+            IUIThreadOperationScope scope = null;
+
+            try
+            {
+                // partial mode is always cancellable
+                var title = string.Format(EditorFeaturesResources.Operation_is_not_ready_for_0_yet_see_task_center_for_more_detail, name);
+                scope = context.AddScope(allowCancellation: true, title);
+
+                var service = solution.Workspace.Services.GetService<ISolutionStatusService>();
+                if (service != null)
+                {
+                    // UIThreadOperation always called from UI thread.
```

Comment:
can we have an assert this is true?

Reply:
I dont have one to assert. editor API will throw, so I think we are fine.

Reply:
> I dont have one to assert.

Could you clarify what that means?  If we expect to be on hte UI thread, there are ways we can assert that expectation.

> editor API will throw,

That may be the behavior now.  But maybe they end up breaking in the future and we don't know it.  If we have a threading invariant that matters, we should test for it.

Reply:
we already have all those kind of thing. I dont believe it will change any. if that's true. all our command handler code should have that as well.